### ISSUE TITLE
Hexdump preview

### ISSUE THREAD

**Issue by ricardo-dias** [ranger/ranger#987]
The goal is to preview the contents of a file hex dump format, with physical offsets, hex bytes, and ascii representation (same output as in xxd).

#### ISSUE TYPE
- Improvement/feature request

#### RUNTIME ENVIRONMENT
- Operating system and version: debian
- Terminal emulator and version: urxvt 9.22
- Python version: 2.7
- Ranger version/commit: 1.9 
- Locale: PT

#### EXPECTED BEHAVIOR
I would expect to preview the contents of a file in hex dump.

#### CURRENT BEHAVIOR
No hex dump preview.

#### CONTEXT
Malware analysis.

**Comment by Vifon** [comment]
It's tough to predict for which types of files the user may want such a preview. I've added the following code to my `scope.sh` to the `handle_mime()` function and it seems to work. Feel free to use it for now (and change the file type if needed):

```
        application/*)
            xxd "${FILE_PATH}" && exit 5
            exit 1;;
```

If you can advise the file types that should use this preview type, I think it can be merged after some adjustments.

Additionally it seems to be pretty slow on larger files. Maybe piping to `head(1)` would help.

**Comment by ricardo-dias** [comment]
Thanks for the fast reply. That works, and the head is also a good idea. But I was thinking more about the capability to trigger the hex preview, for example using a command, instead of depending in the file type. I really like how the preview and pager (rifle) work, so if possible I would like to keep the best of both worlds.

**Comment by Vifon** [comment]
Do you want to it be in the preview pane or in a separate process (via rifle)? The latter should be easy. The former not so much.

**Comment by ricardo-dias** [comment]
Via riffle would be fine too. Can I configure rifle to use xxd if a given command is used?

**Comment by Vifon** [comment]
Yes. You can add it for instance as the last rule. This way it should never be used by default if there is a more specific rule but you can always select it manually after pressing <kbd>r</kbd>.

    else = xxd -- "$1"

**Comment by figbux** [comment]
In  v1.7.1, one can add these lines to `scope.sh` below `case "$mimetype" in`
```
     application/*)                                                              
         try file "$path"                                                        
         try echo                                                                
         try hd "$path" -n 2048 && exit 5                                        
         exit 1;;   
```

**Comment by toonn** [comment]
@figbux, 1.7.1 is really old, consider updating.

Maybe we can include a default for this, for everything `file(1)` detects as being a binary file or for `.o` files or something, @Vifon? Users can still expand the list if they want.

**Comment by vifon** [comment]
> we can include a default for this, for everything file(1) detects as being a binary file or for .o files

Sounds reasonable, yes.

**Comment by vifon** [comment]
…but it doesn't seem to be that simple. I just checked the mime type of my `/bin/fish` and its mime type is `application/x-sharedlib`. We'd really need to use an explicit list of eligible file types, not some more general rules.

**Comment by figbux** [comment]
> We'd really need to use an explicit list of eligible file types, not some more general rules.

Hmm, but I suppose hexdump shold not be something other than a fallback mechanism. There will be always some better parsers for those "unsupported" mime-types.  For example, we can use 'readelf' for linux binaries to show some info. But it'll take time to implement this for all mime types; so until someone specifies the default parser of a mime type, it'd be fine to hexdump.

Also I'm considering plain 'data' files those don't have a specific mime type (or do they?). 'data' is the most worthy type to be hexdumped.

**Comment by toonn** [comment]
@figbux, there's already general fallbacks like `file` if I'm not mistaken. And most users would probably prefer a minorly helpful indication of a file's type over a hexdump they have no idea how to interpret. So I don't think hexdump as a general fallback is feasible.

**Comment by figbux** [comment]
@toonn that's true; I agree. But what will happen to files which `file` says it's `data`?

**Comment by toonn** [comment]
@figbux, do you have an example? There's no such MIME type.

**Comment by figbux** [comment]
Allright! I learned my assumption was wrong:
```
$ dd if=/dev/urandom of=example bs=1M count=1
$ file example
example: data
$ file -i example 
example: application/octet-stream; charset=binary
```
I thought a "`data` " would not have a mime-type but I'm happy I was wrong. It's octet-stream.
